staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -252,16 +256,100 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+				if msgErr != nil {
+					if errors.Is(msgErr, &runtimeCostError{}) {
+						errs = append(errs, field.InternalError(fldPath, msgErr))
+						return errs, -1
+					} else {
+						klog.Warning(msgErr)
+						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+						remainingBudget = newRemainingBudget
+					}
+				} else {
+					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
+					remainingBudget = newRemainingBudget
+				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, fmt.Sprintf("failed rule: %s", ruleErrorString(rule))))
+				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
 			}
 		}
 	}
 	return errs, remainingBudget
 }
 
+const (
+	maxMessageExpressionLen = 5000
+)
+
+type runtimeCostError struct {
+	message string
+}
+
+func (e *runtimeCostError) Error() string {
+	return e.message
+}
+
+func (e *runtimeCostError) Is(err error) bool {
+	_, ok := err.(*runtimeCostError)
+	return ok
+}
+
+// evalMessageExpression evaluates the given message expression and returns the evaluated string form and the remaining budget, or an error if one
+// occurred during evaluation.
+func evalMessageExpression(ctx context.Context, expr celgo.Program, exprSrc string, activation interpreter.Activation, remainingBudget int64) (string, int64, error) {
+	evalResult, evalDetails, err := expr.ContextEval(ctx, activation)
+	if evalDetails == nil {
+		return "", -1, &runtimeCostError{
+			fmt.Sprintf("runtime cost could not be calculated for message expression: %q", exprSrc),
+		}
+	}
+	rtCost := evalDetails.ActualCost()
+	if rtCost == nil {
+		return "", -1, &runtimeCostError{
+			fmt.Sprintf("runtime cost could not be calculated for message expression: %q", exprSrc),
+		}
+	} else if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
+		return "", -1, &runtimeCostError{
+			"messageExpression evaluation failed due to running out of cost budget, no further validation rules will be run",
+		}
+	}
+	if err != nil {
+		if strings.HasPrefix(err.Error(), "operation cancelled: actual cost limit exceeded") {
+			return "", -1, &runtimeCostError{
+				fmt.Sprintf("no further validation rules will be run due to call cost exceeds limit for message expression: %q", exprSrc),
+			}
+		}
+		return "", remainingBudget - int64(*rtCost), fmt.Errorf("message expression evaluation failed due to: %v", err.Error())
+	}
+	messageStr, ok := evalResult.Value().(string)
+	if !ok {
+		return "", remainingBudget - int64(*rtCost), errors.New("message expression failed to convert to string")
+	}
+	trimmedMsgStr := strings.TrimSpace(messageStr)
+	if len(trimmedMsgStr) > maxMessageExpressionLen {
+		return "", remainingBudget - int64(*rtCost), errors.New("message expression generated string beyond allowable length")
+	} else if hasNewlines(trimmedMsgStr) {
+		return "", remainingBudget - int64(*rtCost), errors.New("message expression generated string with newlines")
```

Comment:
The err message should be sync as `messageExpression should not contain line breaks`. 
In the error message, should we use field name `messageExpression`?

Reply:
Right, all error messages should use `messageExpression` and not `message expression` now, fixed